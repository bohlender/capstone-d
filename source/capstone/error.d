module capstone.error;

import std.conv;
import std.exception: enforce, assertNotThrown;
import std.format;
import std.string;

import capstone.internal.api: cs_strerror;

enum ErrorCode {
    Ok = 0,
    OutOfMemory,
    UnsupportedArchitecture,
    InvalidHandle,
    InvalidCshArgument,
    InvalidMode,
    InvalidOption,
    UnavailableInstructionDetail,
    UninitializedDynamicMemoryManagement,
    UnsupportedVersion,
    IrrelevantDataAccessInDietEngine,
    IrrelevantDataAccessInSkipdataMode,
    UnsupportedATnTSyntax,
    UnsupportedIntelSyntax
}

// TODO: Proper error handling
void checkErrorCode(in ErrorCode errno){
    enforce(errno == 0, "Error %s: %s".format(errno.to!int, cs_strerror(errno.to!int).fromStringz));
}

void checkErrorCode(in int errno)
in{
    assertNotThrown!ConvException(errno.to!ErrorCode);
}
body{
    errno.to!ErrorCode.checkErrorCode;
}