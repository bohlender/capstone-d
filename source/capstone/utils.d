module capstone.utils;

import std.meta: AliasSeq, staticMap;
import std.format: format;
import std.exception: enforce;

private enum bool distinctFieldNames(names...) = __traits(compiles,{
    static foreach (name; names)
        static if (is(typeof(name) : string))
            mixin("enum int" ~ name ~ " = 0;");
});

private enum bool distinctTypeNames(names...) = __traits(compiles,{
    static foreach (name; names)
        static if (is(name))
            mixin("enum int" ~ name.mangleof ~ " = 0;");
});

unittest{
	assert(distinctFieldNames!("asdf","asd",int,"asdg"));
	assert(!distinctFieldNames!("asdf","asd","asdf"));
	assert(distinctTypeNames!(int,long,string,"string",uint));
	assert(!distinctTypeNames!(int,long,uint,int));
}

template TaggedUnion(Specs...) if (distinctFieldNames!Specs && distinctTypeNames!Specs){
	template FieldSpec(T, string s){
		alias Type = T;
		alias name = s;
	}
	
	template parseSpecs(Specs...){
		static if(Specs.length==0){
			alias parseSpecs = AliasSeq!();
		}else static if(is(Specs[0])){
			alias parseSpecs = AliasSeq!(FieldSpec!(Specs[0..2]), parseSpecs!(Specs[2..$]));
		}else{
			static assert(0, "Attempted to instantiate TaggedUnion with an invalid argument: " ~ Specs[0].stringof);
		}
		
	}

	alias fieldSpecs = parseSpecs!Specs;
	alias extractType(alias spec) = spec.Type;
	alias extractName(alias spec) = spec.name;
	alias Types = staticMap!(extractType, fieldSpecs);
	alias names = staticMap!(extractName, fieldSpecs);

	struct TaggedUnion{
		import std.variant: Algebraic;
		private Algebraic!Types wrapped;

		// Generate constructors
		static foreach(T; Types){
			this(T val){wrapped = val;}
		}

		// Generate getters
		static foreach(spec; fieldSpecs){
			mixin("@property auto " ~ spec.name ~"() const {
				enforce(wrapped.type == typeid(spec.Type), \"Trying to read a '\" ~ spec.Type.stringof ~ \"' but a '\" ~ wrapped.type.toString ~ \"' is stored\");
				return wrapped.get!(spec.Type);
			}");
		}

		// Generate setters
		static foreach(i, spec; fieldSpecs){
			mixin("@property auto " ~ spec.name ~ "(spec.Type val){wrapped = val;}");
		}
	}
}