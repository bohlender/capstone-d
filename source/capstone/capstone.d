/// Object-oriented wrapper of Capstone disassembly engine
module capstone.capstone;

import std.typecons: Tuple, BitFlags, Yes, Nullable;
import std.format: format;
import std.conv: to;
import std.array: array, appender;
import std.range: isInputRange, enumerate, front;
import std.algorithm: canFind;
import std.traits: EnumMembers;

import capstone.api;
import capstone.instruction;
import capstone.internal;
import capstone.error;

/** Encapsulates an instance of the Capstone dissassembly engine

This class encapsulates the core functionality of the Capstone disassembly engine, providing
access to runtime options for
$(UL
    $(LI changing the `Mode` of interpretation)
    $(LI changing the `Syntax` of the disassembly)
    $(LI choosing whether `Instruction`'s should be disassembled in detail, i.e. filling `Instruction.detail`)
    $(LI defining manual handling of broken instructions through the $(LINK2 http://www.capstone-engine.org/skipdata.html, SKIPDATA) mode of operation (optionally via a `Callback`))
)

Note that, since the architecture is chosen at runtime, this base class only provides access to the architecture-indepentent aspects,
e.g. disasm returns `Instruction`s instead of `X86Instruction`s.
However, if necessary, it can be casted to the architecture-specific variant, such as `CapstoneX86`.
*/
abstract class Capstone{
    package {
        alias Handle = size_t;
        Handle handle;    
        
        ModeFlags _mode;
        Syntax _syntax;
        bool _detail;
        bool _skipData;

        string mnemonic;
        Callback callback;

        string[int] customMnemonics;
    }
    const Arch arch; /// The architecture this Capstone instance is set up for

    /** Constructs an instance of the disassembly engine

    Params:
        arch = The architecture the engine will be created for
        modeFlags = A combination of flags to further specify how bytes will be interpreted, e.g. in little-endian.
    */
    private this(in Arch arch, in ModeFlags modeFlags){
        const libVer = versionOfLibrary;
        const bindVer = versionOfBindings;
        if(libVer != bindVer)
            throw new CapstoneException("API version mismatch between library (%s) and bindings (%s)".format(libVer, bindVer), ErrorCode.UnsupportedVersion);

        // Create Capstone engine instance
        this.arch = arch;
        this._mode = modeFlags;
        cs_open(arch, modeFlags.to!uint, &handle).checkErrno;
    }

    ~this(){
        if(handle)
           cs_close(&handle).checkErrno;
    }
    
    /// Gets the mode of interpretation
    @property auto mode() const {return _mode;}
    /// Sets the mode of interpretation
    @property void mode(in ModeFlags modeFlags){
        _mode = modeFlags;
        cs_option(handle, cs_opt_type.CS_OPT_MODE, modeFlags.to!uint).checkErrno;
    }

    /// Gets the disassembly syntax variant
    @property auto syntax() const {return _syntax;}
    /// Sets the disassembly syntax variant
    @property void syntax(in Syntax option){
        _syntax = option;
        cs_option(handle, cs_opt_type.CS_OPT_SYNTAX, option).checkErrno;
    }

    /// Indicates whether instructions will be disassembled in detail
    @property auto detail() const {return _detail;}
    /// Sets whether instructions will be disassembled in detail
    @property void detail(in bool enable){
        _detail = enable;
        auto option = enable ? cs_opt_value.CS_OPT_ON : cs_opt_value.CS_OPT_OFF;
        cs_option(handle, cs_opt_type.CS_OPT_DETAIL, option).checkErrno;
    }

    /// Indicates whether SKIPDATA mode of operation is in use
    @property auto skipData() const {return _skipData;}
    /// Sets whether to use SKIPDATA mode of operation
    @property void skipData(in bool enable){
        _skipData = enable;
        auto option = enable ? cs_opt_value.CS_OPT_ON : cs_opt_value.CS_OPT_OFF;
        cs_option(handle, cs_opt_type.CS_OPT_SKIPDATA, option).checkErrno;
    }

    /** Customises behaviour in SKIPDATA mode of operation
     
    By default, disassembling will stop when it encounters a broken instruction.
    Most of the time, the reason is that this is data mixed inside the input.

    When in SKIPDATA mode, some (unknown) amount of data until the next interpretable instruction will be skipped.
    Capstone considers the skipped data a special instruction with ID 0x00 and a `mnemonic` that defaults to `".byte"`.
    The operand string is a hex-code of the sequence of bytes it skipped.

    By default, for each iteration, Capstone skips 1 byte on X86 architecture, 2 bytes on Thumb mode on Arm
    architecture, and 4 bytes for the rest. The reason while Capstone skips 1 byte on X86 is that X86 puts no
    restriction on instruction alignment, but other architectures enforce some requirements on this aspect.

    To customise how many bytes to skip when encountering data, a `Callback` delegate can optonally be setup
    to return the corresponding number.

    Params:
        mnemonic = The mnemonic to use for representing skipped data
        callback = The optional callback to use for handling bytes that cannot be interpreted as an instruction.
    
    Example:
    ---
    // Custom data that can be referred to in a callback delegate
    struct CallbackData{
        int bytesToSkip;
    }
    auto myData = CallbackData(1);

    size_t myCallback(in ubyte[] code, size_t offset) {
        return myData.bytesToSkip++; // Always skip one more byte when encountering data
    }

    cs.skipData = true;                     // Enable skipdata mode
    cs.setupSkipdata("db", &myCallback);    // Use custom callback, and "db" as custom mnemonic for data
    ---
    */
    void setupSkipdata(in string mnemonic = ".byte", Callback callback = null){
        if(!mnemonic)
            throw new CapstoneException("Invalid mnemonic", ErrorCode.InvalidOption);
        this.mnemonic = mnemonic;
        this.callback = callback;
        
        auto setup = cs_opt_skipdata(this.mnemonic.ptr, this.callback ? &cCallback : null, &this.callback);
        cs_option(handle, cs_opt_type.CS_OPT_SKIPDATA_SETUP, cast(size_t)&setup).checkErrno;
    }

    /// Defines a custom mnemonic for a specified instruction id
    private void customMnemonic(in int id, in string mnem = null) {
        auto optMnem = cs_opt_mnem(id, null);
        if(mnem != null){
            const v = (customMnemonics[id] = mnem);
            optMnem.mnemonic = v.ptr;
        }else
            customMnemonics.remove(id);
        cs_option(handle, cs_opt_type.CS_OPT_MNEMONIC, cast(size_t)&optMnem).checkErrno;
    }

    /** Disassemble binary code, given the code buffer, start address and number of instructions to be decoded
    
    Params:
        code    = Buffer containing raw binary code to be disassembled
        address = Address of the first instruction in given raw code buffer
        count   = Number of instructions to be disassembled, or 0 to get all of them
    Returns: The successfully disassembled instructions

    Example:
    ---
    auto CODE = cast(ubyte[])"\x8d\x4c\x32\x08\x01\xd8\x81\xc6\x34\x12\x00\x00\x00\x91\x92";
    auto cs = new CapstoneX86(ModeFlags(Mode.bit32)); // Initialise x86 32bit engine
    auto res = cs.disasm(CODE, 0x1000);               // Disassemble, offsetting addresses by 0x1000
    assert("%s %s".format(res[0].mnemonic, res[0].opStr) == "lea ecx, dword ptr [edx + esi + 8]");
    assert("%s %s".format(res[1].mnemonic, res[1].opStr) == "add eax, ebx");
    assert("%s %s".format(res[2].mnemonic, res[2].opStr) == "add esi, 0x1234");
    ---
    */
    abstract const(Instruction)[] disasm(in ubyte[] code, in ulong address, in size_t count = 0) const;

    /** Provides a range to iteratively disassemble binary code - one instruction at a time

    Fast API to disassemble binary code, given the code buffer and start address.
    Provides access to only one disassembled instruction at a time, resulting in a smaller memory footprint.
    Params:
        code    = Buffer containing raw binary code to be disassembled
        address = Address of the first instruction in given raw code buffer
    Returns: An input range over the disassembled instructions
    Example:
    ---
    auto CODE = cast(ubyte[])"\x8d\x4c\x32\x08\x01\xd8\x81\xc6\x34\x12\x00\x00\x00\x91\x92";
    auto cs = new CapstoneX86(ModeFlags(Mode.bit32)); // Initialise x86 32bit engine
    auto range = cs.disasmIter(CODE, 0x1000);         // Disassemble one instruction at a time, offsetting addresses by 0x1000
    assert("%s %s".format(range.front.mnemonic, range.front.opStr) == "lea ecx, dword ptr [edx + esi + 8]");
    range.popFront;
    assert("%s %s".format(range.front.mnemonic, range.front.opStr) == "add eax, ebx");
    range.popFront;
    assert("%s %s".format(range.front.mnemonic, range.front.opStr) == "add esi, 0x1234");
    range.popFront;
    assert(range.empty);
    ---
    */
    abstract InstructionRange disasmIter(in ubyte[] code, in ulong address) const;
}

// TODO: Try switching to InputRange!Instruction (more restrictive than isInputRange, though)
/// An input range that provides access to one disassembled `Instruction` at a time
abstract class InstructionRange {
    /// Retrieves element of the range
    Instruction front();
    /// True if range has no instructions, i.e. cannot be advanced anymore
    bool empty();
    /// Drops the front instruction and advances the range
    void popFront();
}
static assert(isInputRange!InstructionRange);

unittest{ // disasm
    auto CODE = cast(ubyte[])"\x8d\x4c\x32\x08\x01\xd8\x81\xc6\x34\x12\x00\x00\x00\x91\x92";
    auto cs = create(Arch.x86, ModeFlags(Mode.bit16));
    cs.mode = ModeFlags(Mode.bit32);

    auto res = cs.disasm(CODE, 0x1000);
    assert(res.length == 3); // With skipdata disabled, disassembling will halt when encountering data
    assert("%s %s".format(res[0].mnemonic, res[0].opStr) == "lea ecx, [edx + esi + 8]");
    assert("%s %s".format(res[1].mnemonic, res[1].opStr) == "add eax, ebx");
    assert("%s %s".format(res[2].mnemonic, res[2].opStr) == "add esi, 0x1234");
    cs.skipData = true;
    res = cs.disasm(CODE, 0x1000, 5);
    assert(res.length == 5);
}

/** Class template that encapsulates an architecture-specific instance of the Capstone dissassembly engine

Note that all architecture-specific instances, like `CapstoneX86`, instantiate and derive from this one.

Params:
    TInstructionId = The architecture-specific instruction identifier type
    TInstruction = The architecture-specific instruction type
*/
abstract class CapstoneImpl(TInstructionId, TInstruction) : Capstone {
    import capstone.range: InstructionImplRange;

    /** Creates an architecture-specific instance with a given mode of interpretation
    
    Params:
        arch = The architecture the engine will be created for
        modeFlags = The (initial) mode of interpretation, which can still be changed later on
    */
    package this(in Arch arch, in ModeFlags modeFlags){
        super(arch, modeFlags);
    }

    override TInstruction[] disasm(in ubyte[] code, in ulong address, in size_t count = 0) const {
        auto instrAppnd = appender!(TInstruction[]);
        foreach(i, instr; disasmIter(code, address).enumerate){
            instrAppnd.put(instr);
            if(i+1==count)
                break;
        }
        return instrAppnd.data;
    }

    override InstructionImplRange!TInstruction disasmIter(in ubyte[] code, in ulong address) const {
        return new InstructionImplRange!TInstruction(this, code, address);
    }

    /** Defines a custom mnemonic for a specified instruction id
    
    Example:
    ---
    enum X86_CODE32 = cast(ubyte[])"\x75\x01";
    auto cs = new CapstoneX86(ModeFlags(Mode.bit32));
    
    // Customize mnemonic JNE to JNZ
    cs.customMnemonic(X86InstructionId.jne, "jnz");
    
    foreach(instr; cs.disasm(X86_CODE32, 0x1000))
        writefln("%s\t%s", instr.mnemonic, instr.opStr);
    
    // Reset engine to use the default mnemonic of JNE
    cs.customMnemonic(X86InstructionId.jne);
    
    foreach(instr; cs.disasm(X86_CODE32, 0x1000))
        writefln("%s\t%s", instr.mnemonic, instr.opStr);
    ---
    */
    void customMnemonic(in TInstructionId id, in string mnem = null) {
        super.customMnemonic(id, mnem);
    }
}