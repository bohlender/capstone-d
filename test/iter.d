module test.iter;

import std.outbuffer;

import capstone;
import test.utils;

enum X86_CODE16 = cast(ubyte[])"\x8d\x4c\x32\x08\x01\xd8\x81\xc6\x34\x12\x00\x00";
enum X86_CODE32 = cast(ubyte[])"\x8d\x4c\x32\x08\x01\xd8\x81\xc6\x34\x12\x00\x00";
enum X86_CODE64 = cast(ubyte[])"\x55\x48\x8b\x05\xb8\x13\x00\x00";
enum ARM_CODE = cast(ubyte[])"\xED\xFF\xFF\xEB\x04\xe0\x2d\xe5\x00\x00\x00\x00\xe0\x83\x22\xe5\xf1\x02\x03\x0e\x00\x00\xa0\xe3\x02\x30\xc1\xe7\x00\x00\x53\xe3";
enum ARM_CODE2 = cast(ubyte[])"\x10\xf1\x10\xe7\x11\xf2\x31\xe7\xdc\xa1\x2e\xf3\xe8\x4e\x62\xf3";
enum THUMB_CODE = cast(ubyte[])"\x70\x47\xeb\x46\x83\xb0\xc9\x68";
enum THUMB_CODE2 = cast(ubyte[])"\x4f\xf0\x00\x01\xbd\xe8\x00\x88\xd1\xe8\x00\xf0";
enum MIPS_CODE = cast(ubyte[])"\x0C\x10\x00\x97\x00\x00\x00\x00\x24\x02\x00\x0c\x8f\xa2\x00\x00\x34\x21\x34\x56\x00\x80\x04\x08";
enum MIPS_CODE2 = cast(ubyte[])"\x56\x34\x21\x34\xc2\x17\x01\x00";
enum ARM64_CODE = cast(ubyte[])"\x09\x00\x38\xd5\xbf\x40\x00\xd5\x0c\x05\x13\xd5\x20\x50\x02\x0e\x20\xe4\x3d\x0f\x00\x18\xa0\x5f\xa2\x00\xae\x9e\x9f\x37\x03\xd5\xbf\x33\x03\xd5\xdf\x3f\x03\xd5\x21\x7c\x02\x9b\x21\x7c\x00\x53\x00\x40\x21\x4b\xe1\x0b\x40\xb9\x20\x04\x81\xda\x20\x08\x02\x8b\x10\x5b\xe8\x3c";
enum PPC_CODE = cast(ubyte[])"\x80\x20\x00\x00\x80\x3f\x00\x00\x10\x43\x23\x0e\xd0\x44\x00\x80\x4c\x43\x22\x02\x2d\x03\x00\x80\x7c\x43\x20\x14\x7c\x43\x20\x93\x4f\x20\x00\x21\x4c\xc8\x00\x21\x40\x82\x00\x14";
enum SPARC_CODE = cast(ubyte[])"\x80\xa0\x40\x02\x85\xc2\x60\x08\x85\xe8\x20\x01\x81\xe8\x00\x00\x90\x10\x20\x01\xd5\xf6\x10\x16\x21\x00\x00\x0a\x86\x00\x40\x02\x01\x00\x00\x00\x12\xbf\xff\xff\x10\xbf\xff\xff\xa0\x02\x00\x09\x0d\xbf\xff\xff\xd4\x20\x60\x00\xd4\x4e\x00\x16\x2a\xc2\x80\x03";
enum SPARCV9_CODE = cast(ubyte[])"\x81\xa8\x0a\x24\x89\xa0\x10\x20\x89\xa0\x1a\x60\x89\xa0\x00\xe0";
enum SYSZ_CODE = cast(ubyte[])"\xed\x00\x00\x00\x00\x1a\x5a\x0f\x1f\xff\xc2\x09\x80\x00\x00\x00\x07\xf7\xeb\x2a\xff\xff\x7f\x57\xe3\x01\xff\xff\x7f\x57\xeb\x00\xf0\x00\x00\x24\xb2\x4f\x00\x78";
enum XCORE_CODE = cast(ubyte[])"\xfe\x0f\xfe\x17\x13\x17\xc6\xfe\xec\x17\x97\xf8\xec\x4f\x1f\xfd\xec\x37\x07\xf2\x45\x5b\xf9\xfa\x02\x06\x1b\x10";

// TODO: Add missing tests when all archs are supported
/*
enum platforms = [
	Platform(Arch.x86, Mode.bit16, X86_CODE16, "X86 16bit (Intel syntax)"),
	Platform(Arch.x86, Mode.bit32, X86_CODE32, "X86 32bit (ATT syntax)", Syntax.att),
	Platform(Arch.x86, Mode.bit32, X86_CODE32, "X86 32 (Intel syntax)"),
	Platform(Arch.x86, Mode.bit64, X86_CODE64, "X86 64 (Intel syntax)"),
	Platform(Arch.arm, Mode.arm, ARM_CODE, "ARM"),
	Platform(Arch.arm, Mode.armThumb, THUMB_CODE2, "THUMB-2"),
	Platform(Arch.arm, Mode.arm, ARM_CODE2, "ARM: Cortex-A15 + NEON"),
	Platform(Arch.arm, Mode.armThumb, THUMB_CODE, "THUMB"),
	Platform(Arch.mips, Mode.mips32 + Mode.bigEndian, MIPS_CODE, "MIPS-32 (Big-endian)"),
	Platform(Arch.mips, Mode.mips64 + Mode.littleEndian, MIPS_CODE2, "MIPS-64-EL (Little-endian)"),
	Platform(Arch.arm64, Mode.arm, ARM64_CODE, "ARM-64"),
	Platform(Arch.powerPc, Mode.bigEndian, PPC_CODE, "PPC-64"),
	Platform(Arch.sparc, Mode.bigEndian, SPARC_CODE, "Sparc"),
	Platform(Arch.sparc, Mode.bigEndian + Mode.sparcV9, SPARCV9_CODE, "SparcV9"),
	Platform(Arch.systemZ, Mode.littleEndian, SYSZ_CODE, "SystemZ"),
	Platform(Arch.xCore, Mode.littleEndian, XCORE_CODE, "XCore")
]*/
enum platforms = [
	Platform(Arch.x86, Mode.bit16, X86_CODE16, "X86 16bit (Intel syntax)"),
	Platform(Arch.x86, Mode.bit32, X86_CODE32, "X86 32bit (ATT syntax)", Syntax.att),
	Platform(Arch.x86, Mode.bit32, X86_CODE32, "X86 32 (Intel syntax)"),
	Platform(Arch.x86, Mode.bit64, X86_CODE64, "X86 64 (Intel syntax)"),
	Platform(Arch.arm, Mode.arm, ARM_CODE, "ARM"),
	Platform(Arch.arm, Mode.armThumb, THUMB_CODE2, "THUMB-2"),
	Platform(Arch.arm, Mode.arm, ARM_CODE2, "ARM: Cortex-A15 + NEON"),
	Platform(Arch.arm, Mode.armThumb, THUMB_CODE, "THUMB"),
	Platform(Arch.mips, Mode.mips32 + Mode.bigEndian, MIPS_CODE, "MIPS-32 (Big-endian)"),
	Platform(Arch.mips, Mode.mips64 + Mode.littleEndian, MIPS_CODE2, "MIPS-64-EL (Little-endian)"),
	Platform(Arch.arm64, Mode.arm, ARM64_CODE, "ARM-64"),
	Platform(Arch.ppc, Mode.bigEndian, PPC_CODE, "PPC-64"),
	Platform(Arch.sparc, Mode.bigEndian, SPARC_CODE, "Sparc"),
	Platform(Arch.sparc, Mode.bigEndian + Mode.sparcV9, SPARCV9_CODE, "SparcV9"),
];

void writeDetail(Arch arch)(ref OutBuffer buf, in InstructionImpl!arch instr, in CapstoneImpl!arch cs) {
	buf.writefln("0x%x:\t%s\t\t%s // insn-ID: %d, insn-mnem: %s", instr.address, instr.mnemonic, instr.opStr, instr.id, cs.instrName(instr.id));
	auto detail = instr.detail;
	if(detail.regsRead.length > 0) {
		buf.writef("\tImplicit registers read: ");
		foreach(reg; detail.regsRead)
			buf.writef("%s ", cs.regName(reg));
		buf.writefln("");
	}
	if(detail.regsWrite.length > 0) {
		buf.writef("\tImplicit registers modified: ");
		foreach(reg; detail.regsWrite)
			buf.writef("%s ", cs.regName(reg));
		buf.writefln("");
	}
	if(detail.groups.length > 0) {
		buf.writef("\tThis instruction belongs to groups: ");
		foreach(group; detail.groups)
			buf.writef("%s ", cs.groupName(group));
		buf.writefln("");
	}
}

unittest{
	auto buf = new OutBuffer;
	foreach(i, platform; platforms) {
		auto cs = Capstone.create(platform.arch, ModeFlags(platform.mode));
		cs.syntax = platform.syntax;
		cs.detail = true;

		buf.writefln("****************");
		buf.writefln("Platform: %s", platform.comment);
		buf.writefln("Code: %s", platform.code.bytesToHex);
		buf.writefln("Disasm:");

		foreach(instr; cs.disasmIter(platform.code, 0x1000)){ 
			switch(platform.arch){
				case Arch.arm:
					buf.writeDetail(cast(InstructionImpl!(Arch.arm))instr, cast(CapstoneImpl!(Arch.arm))cs);
					break;
				case Arch.arm64:
					buf.writeDetail(cast(InstructionImpl!(Arch.arm64))instr, cast(CapstoneImpl!(Arch.arm64))cs);
					break;
				case Arch.mips:
					buf.writeDetail(cast(InstructionImpl!(Arch.mips))instr, cast(CapstoneImpl!(Arch.mips))cs);
					break;
				case Arch.ppc:
					buf.writeDetail(cast(InstructionImpl!(Arch.ppc))instr, cast(CapstoneImpl!(Arch.ppc))cs);
					break;
				case Arch.sparc:
					buf.writeDetail(cast(InstructionImpl!(Arch.sparc))instr, cast(CapstoneImpl!(Arch.sparc))cs);
					break;
				case Arch.x86:
					buf.writeDetail(cast(InstructionImpl!(Arch.x86))instr, cast(CapstoneImpl!(Arch.x86))cs);
					break;
				default:
					assert(false);
			}
		}
		buf.writefln("");
	}

	const expected = import("iter.expected");
	const actual = buf.toString;

	assert(expected == actual, expectationMismatch(expected, actual));
}