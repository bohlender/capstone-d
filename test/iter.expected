****************
Platform: X86 16bit (Intel syntax)
Code: 0x8d 0x4c 0x32 0x08 0x01 0xd8 0x81 0xc6 0x34 0x12 0x00 0x00 
Disasm:
0x1000:	lea		cx, word ptr [si + 0x32] // insn-ID: 315, insn-mnem: lea
0x1003:	or		byte ptr [bx + di], al // insn-ID: 325, insn-mnem: or
	Implicit registers modified: flags 
0x1005:	fadd		dword ptr [bx + di + 0x34c6] // insn-ID: 15, insn-mnem: fadd
	Implicit registers modified: fpsw 
0x1009:	adc		al, byte ptr [bx + si] // insn-ID: 6, insn-mnem: adc
	Implicit registers read: flags 
	Implicit registers modified: flags 

****************
Platform: X86 32bit (ATT syntax)
Code: 0x8d 0x4c 0x32 0x08 0x01 0xd8 0x81 0xc6 0x34 0x12 0x00 0x00 
Disasm:
0x1000:	leal		8(%edx, %esi), %ecx // insn-ID: 315, insn-mnem: lea
	This instruction belongs to groups: not64bitmode 
0x1004:	addl		%ebx, %eax // insn-ID: 8, insn-mnem: add
	Implicit registers modified: eflags 
0x1006:	addl		$0x1234, %esi // insn-ID: 8, insn-mnem: add
	Implicit registers modified: eflags 

****************
Platform: X86 32 (Intel syntax)
Code: 0x8d 0x4c 0x32 0x08 0x01 0xd8 0x81 0xc6 0x34 0x12 0x00 0x00 
Disasm:
0x1000:	lea		ecx, dword ptr [edx + esi + 8] // insn-ID: 315, insn-mnem: lea
	This instruction belongs to groups: not64bitmode 
0x1004:	add		eax, ebx // insn-ID: 8, insn-mnem: add
	Implicit registers modified: eflags 
0x1006:	add		esi, 0x1234 // insn-ID: 8, insn-mnem: add
	Implicit registers modified: eflags 

****************
Platform: X86 64 (Intel syntax)
Code: 0x55 0x48 0x8b 0x05 0xb8 0x13 0x00 0x00 
Disasm:
0x1000:	push		rbp // insn-ID: 580, insn-mnem: push
	Implicit registers read: rsp 
	Implicit registers modified: rsp 
	This instruction belongs to groups: mode64 
0x1001:	mov		rax, qword ptr [rip + 0x13b8] // insn-ID: 442, insn-mnem: mov

****************
Platform: ARM
Code: 0xed 0xff 0xff 0xeb 0x04 0xe0 0x2d 0xe5 0x00 0x00 0x00 0x00 0xe0 0x83 0x22 0xe5 0xf1 0x02 0x03 0x0e 0x00 0x00 0xa0 0xe3 0x02 0x30 0xc1 0xe7 0x00 0x00 0x53 0xe3 
Disasm:
0x1000:	bl		#0xfbc // insn-ID: 13, insn-mnem: bl
	Implicit registers read: pc 
	Implicit registers modified: lr 
	This instruction belongs to groups: arm jump 
0x1004:	str		lr, [sp, #-4]! // insn-ID: 212, insn-mnem: str
	This instruction belongs to groups: arm 
0x1008:	andeq		r0, r0, r0 // insn-ID: 8, insn-mnem: and
	This instruction belongs to groups: arm 
0x100c:	str		r8, [r2, #-0x3e0]! // insn-ID: 212, insn-mnem: str
	This instruction belongs to groups: arm 
0x1010:	mcreq		p2, #0, r0, c3, c1, #7 // insn-ID: 74, insn-mnem: mcr
	This instruction belongs to groups: arm 
0x1014:	mov		r0, #0 // insn-ID: 80, insn-mnem: mov
	This instruction belongs to groups: arm 
0x1018:	strb		r3, [r1, r2] // insn-ID: 203, insn-mnem: strb
	This instruction belongs to groups: arm 
0x101c:	cmp		r3, #0 // insn-ID: 23, insn-mnem: cmp
	Implicit registers modified: cpsr 
	This instruction belongs to groups: arm 

****************
Platform: THUMB-2
Code: 0x4f 0xf0 0x00 0x01 0xbd 0xe8 0x00 0x88 0xd1 0xe8 0x00 0xf0 
Disasm:
0x1000:	mov.w		r1, #0 // insn-ID: 80, insn-mnem: mov
	This instruction belongs to groups: thumb2 
0x1004:	pop.w		{fp, pc} // insn-ID: 425, insn-mnem: pop
	This instruction belongs to groups: thumb2 
0x1008:	tbb		[r1, r0] // insn-ID: 420, insn-mnem: tbb
	This instruction belongs to groups: thumb2 jump 

****************
Platform: ARM: Cortex-A15 + NEON
Code: 0x10 0xf1 0x10 0xe7 0x11 0xf2 0x31 0xe7 0xdc 0xa1 0x2e 0xf3 0xe8 0x4e 0x62 0xf3 
Disasm:
0x1000:	sdiv		r0, r0, r1 // insn-ID: 122, insn-mnem: sdiv
	This instruction belongs to groups: arm 
0x1004:	udiv		r1, r1, r2 // insn-ID: 231, insn-mnem: udiv
	This instruction belongs to groups: arm 
0x1008:	vbit		q5, q15, q6 // insn-ID: 274, insn-mnem: vbit
	This instruction belongs to groups: neon 
0x100c:	vcgt.f32		q10, q9, q12 // insn-ID: 278, insn-mnem: vcgt
	This instruction belongs to groups: neon 

****************
Platform: THUMB
Code: 0x70 0x47 0xeb 0x46 0x83 0xb0 0xc9 0x68 
Disasm:
0x1000:	bx		lr // insn-ID: 15, insn-mnem: bx
	This instruction belongs to groups: thumb jump 
0x1002:	mov		fp, sp // insn-ID: 80, insn-mnem: mov
	This instruction belongs to groups: thumb thumb1only 
0x1004:	sub		sp, #0xc // insn-ID: 213, insn-mnem: sub
	This instruction belongs to groups: thumb thumb1only 
0x1006:	ldr		r1, [r1, #0xc] // insn-ID: 73, insn-mnem: ldr
	This instruction belongs to groups: thumb thumb1only 

****************
Platform: ARM-64
Code: 0x09 0x00 0x38 0xd5 0xbf 0x40 0x00 0xd5 0x0c 0x05 0x13 0xd5 0x20 0x50 0x02 0x0e 0x20 0xe4 0x3d 0x0f 0x00 0x18 0xa0 0x5f 0xa2 0x00 0xae 0x9e 0x9f 0x37 0x03 0xd5 0xbf 0x33 0x03 0xd5 0xdf 0x3f 0x03 0xd5 0x21 0x7c 0x02 0x9b 0x21 0x7c 0x00 0x53 0x00 0x40 0x21 0x4b 0xe1 0x0b 0x40 0xb9 0x20 0x04 0x81 0xda 0x20 0x08 0x02 0x8b 0x10 0x5b 0xe8 0x3c 
Disasm:
0x1000:	mrs		x9, midr_el1 // insn-ID: 192, insn-mnem: mrs
0x1004:	msr		spsel, #0 // insn-ID: 193, insn-mnem: msr
	Implicit registers modified: nzcv 
0x1008:	msr		dbgdtrtx_el0, x12 // insn-ID: 193, insn-mnem: msr
0x100c:	tbx		v0.8b, {v1.16b, v2.16b, v3.16b}, v2.8b // insn-ID: 347, insn-mnem: tbx
	This instruction belongs to groups: neon 
0x1010:	scvtf		v0.2s, v1.2s, #3 // insn-ID: 234, insn-mnem: scvtf
	This instruction belongs to groups: neon 
0x1014:	fmla		s0, s0, v0.s[3] // insn-ID: 114, insn-mnem: fmla
	This instruction belongs to groups: neon 
0x1018:	fmov		x2, v5.d[1] // insn-ID: 116, insn-mnem: fmov
	This instruction belongs to groups: fparmv8 
0x101c:	dsb		nsh // insn-ID: 60, insn-mnem: dsb
0x1020:	dmb		osh // insn-ID: 58, insn-mnem: dmb
0x1024:	isb		 // insn-ID: 142, insn-mnem: isb
0x1028:	mul		x1, x1, x2 // insn-ID: 195, insn-mnem: mul
0x102c:	lsr		w1, w1, #0 // insn-ID: 184, insn-mnem: lsr
0x1030:	sub		w0, w0, w1, uxtw // insn-ID: 340, insn-mnem: sub
0x1034:	ldr		w1, [sp, #8] // insn-ID: 162, insn-mnem: ldr
0x1038:	cneg		x0, x1, ne // insn-ID: 440, insn-mnem: cneg
	Implicit registers read: nzcv 
0x103c:	add		x0, x1, x2, lsl #2 // insn-ID: 6, insn-mnem: add
0x1040:	ldr		q16, [x24, w8, uxtw #4] // insn-ID: 162, insn-mnem: ldr

